### PR TITLE
add Rate Tap button

### PR DESCRIPTION
**Pull request by ronso0**
* add `tempo_tap` which works exactly like BPM tap, just it sets the tempo
* change BPM display overlay and buttons in skins to do
  * **tempo** tap on left click
  * BPM tap on right-click.

quoting myself from #12086
> Admittedly, I never understood why this a track edit tool [BPM tap button] was placed in the rate control section in the first place.

> double-clicking any track (text) label opens the Track Properties editor where the (track) BPM can be set (type or tap)
Use case

### Reasoning
BPM tap edits a track property (permanently) while the rate section is solely about controlling the playback speed (+, -, sync). Hence, IMO BPM tap doesn't belong there.
IIUC the intention for putting it there was to allow _setting_ the track BPM, i.e. set initial value or correct BPM analysis. I assume it was helpful in the early days of Mixxx but with the improved beats analyzer nowadays I suppose it's not needed (that much) anymore. Manual correction can be done in Track Properties > BPM tab.

### My use case
Infrequently I need to "Sync" to external players (vinyl via Aux, other DJ in b2b sessions), i.e. at least estimate the target BPM. Until now I was using a Python BPM counter [1], activated with a global hotkey, which gives very accurate results. However, it'd be much more convenient to do this within Mixxx.
Either I apply the guessed target BPM (e.g. 128) to loaded tracks, or I do a bpm search `bpm:125-130`.
Rate Tap is a good step forward.

Thinking this further, a BPM tap utility might be helpful, see #12105 

- 1 https://github.com/m13253/bpm-counter/blob/master/trunk/bpm-counter.py

### CODE REVIEW

**Comment by deejay-girl** [comment]
Where can we download nightly build of Mixxx?
The 2.6 alpha development version on the website is never updated so for me it's impossible to test the latest changes :(
I can't compile Mixxx myself.

**Comment by ronso0** [comment]
Getting alpha builds and testing Pull Requests is described on the wiki
https://github.com/mixxxdj/mixxx/wiki/Testing

**Comment by ronso0** [comment]
Very weird: I built this locally and it works as expected.
Though, it was built with Qt5 because I forgot to set the Qt6 switch.
So with Qt6 [`slotUpdateRateSlider(double bpm)`](https://github.com/mixxxdj/mixxx/blob/0c2faf06a89e5aca14d1aa7df8e90f9c5b50c69c/src/engine/controls/bpmcontrol.cpp#L939) is not called when BPM are tapped, even though it should
https://github.com/mixxxdj/mixxx/blob/0c2faf06a89e5aca14d1aa7df8e90f9c5b50c69c/src/engine/controls/bpmcontrol.cpp#L101-L111

When I change `bpm` from dev tools it works, it just doesn't when `bpm` is set in `slotRateTapFilter` :shrug: 
It also works if I make `bpm` read-only and use `connectValueChangeRequest` and copy all required calls there.

Also, I have no idea why unrelated tests fail n all OS except ARM 64, for example https://github.com/mixxxdj/mixxx/actions/runs/6517384471/job/17707390983?pr=12104

I'd appreciate if someone could give me some hints what's wrong here.

**Comment by ronso0** [comment]
I pushed a different (working) version with my findings added inline.

**Comment by ronso0** [comment]
> Very weird: I built this locally and it works as expected. Though, it was built with Qt5 because I forgot to set the Qt6 switch. So with Qt6 [`slotUpdateRateSlider(double bpm)`](https://github.com/mixxxdj/mixxx/blob/0c2faf06a89e5aca14d1aa7df8e90f9c5b50c69c/src/engine/controls/bpmcontrol.cpp#L939) is not called when BPM are tapped, even though it should
...

Seems this is a known issue? Just stumbled upon https://github.com/mixxxdj/mixxx/blob/558253682acc5c07452f829209fefca91bd8ba8b/src/broadcast/broadcastmanager.cpp#L71-L78
Still wondering why the comment was written for Qt5 whereas it worked here with Qt5 and fails with Qt6 :shrug: 
I changed it to call the slot manually.

**Comment by ronso0** [comment]
I changed all skin buttons to tap the tempo on left click and tap the BPM on right click.

**Comment by ronso0** [comment]
I changed it to `tempo_tap` to make it more clear what it does.

Ready to roll!

File: src/engine/controls/bpmcontrol.h [hunk 1]
```diff
@@ -123,51 +124,53 @@ class BpmControl : public EngineControl {
     friend class SyncControl;
 
     // ControlObjects that come from EngineBuffer
-    ControlProxy* m_pPlayButton;
+    std::unique_ptr<ControlProxy> m_pPlayButton;
     QAtomicInt m_oldPlayButton;
-    ControlProxy* m_pReverseButton;
-    ControlProxy* m_pRateRatio;
+    std::unique_ptr<ControlProxy> m_pReverseButton;
```
Comment (ronso0) [inline src/engine/controls/bpmcontrol.h:129]: note: can be PollingControlProxy

**Comment by fwcd** [comment]
> I assume it was helpful in the early days of Mixxx but with the improved beats analyzer nowadays I suppose it's not needed (that much) anymore

I use the BPM tap pretty frequently on tracks where the analyzer still struggles (acapellas or tracks with complex polyrhythms, e.g. Salsa music). Would appreciate to at least keep the option to keep having BPM tap available in the UI, having to dig into the track properties makes this too inconvenient to really be useful IMHO.

**Comment by ronso0** [comment]
> Moving the BPM tap to the beatgrid editing sections seems reasonable too, to untangle the BPM editing controls from rate changing.

Good idea!
Feasible in Tango and LateNight, not so nice in Deere but the BPM control section is already interfering anyway with the track dsiplays.

> perhaps a switch in the skin settings to toggle left/right-click behavior?

Hmm, the skin settings menu is growing and growing, in LateNight we already had to squeeze so it fits at minimum height.
At some point we need to wrap it into a WScrollable #3890 (still unused in official skins)

**Comment by daschuer** [comment]
Can you move the first commit to a separate PR? 
This way we see here better the proposed functional changes.

**Comment by ronso0** [comment]
Why not view by commit?

**Comment by daschuer** [comment]
Now a conflict has developed.

**Comment by ronso0** [comment]
> > Moving the BPM tap to the beatgrid editing sections seems reasonable too, to untangle the BPM editing controls from rate changing.

Just filed a feature request to track this.

> > perhaps a switch in the skin settings to toggle left/right-click behavior?
> 
> Hmm, the skin settings menu is growing and growing, in LateNight we already had to squeeze so it fits at minimum height. At some point we need to wrap it into a WScrollable #3890 (still unused in official skins)

I'll look into this during beta.

**Comment by ronso0** [comment]
Conflicts are resolved.

**Comment by daschuer** [comment]
Now the unit-test are failing.

File: src/engine/controls/bpmcontrol.cpp [hunk 2]
```diff
@@ -99,61 +123,61 @@ BpmControl::BpmControl(const QString& group,
...
+    connect(&m_tempoTapFilter,
+            &TapFilter::tapped,
+            this,
+            &BpmControl::slotTempoTapFilter,
             Qt::DirectConnection);
 
-    m_pBpmLock = std::make_unique<ControlPushButton>(
-            ConfigKey(group, "bpmlock"), false);
-    m_pBpmLock->setButtonMode(ControlPushButton::TOGGLE);
-    m_pBpmLock->connectValueChangeRequest(
```
Comment (daschuer) [inline src/engine/controls/bpmcontrol.cpp:136]: because of this change m_pBpmLock is left null and crashes when used in trackBpmLockChanged

**Review by daschuer — CHANGES_REQUESTED** [review]

**Comment by daschuer** [comment]
Oh sorry, and new conflicts have developed.

**Comment by ronso0** [comment]
> Oh sorry, and new conflicts have developed.

Yeah, a lot has happend in BpmControl since I started this.
I rebased and it looks & works good now, let's see what CI says.

**Comment by ronso0** [comment]
small TODO:
highlight the tap button also on right-click. Not sure how to do that, though.

File: src/engine/controls/bpmcontrol.cpp [hunk 3]
```diff
@@ -1007,15 +1007,14 @@ mixxx::audio::FrameDiff_t BpmControl::getPhaseOffset(mixxx::audio::FramePos this
 
 void BpmControl::slotUpdateEngineBpm(double value) {
     // Adjust playback bpm in response to a rate_ratio update
-    double dRate = m_pRateRatio->get();
```
Comment (Swiftb0y) [inline src/engine/controls/bpmcontrol.cpp:1010]: please update the name `value` to reflect the fact that its now the `rate`.

**Review by Swiftb0y — APPROVED** [review]
Only nitpick. rest LGTM. haven't tested though

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
Thank you.

**Pull request merged**